### PR TITLE
2-2 jobs skill rebalance - Sage (2018 patch/Renewal)

### PR DESCRIPTION
**Pull request by guilherme-gm**
### Pull Request Prelude
- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
The main purpose of this PR is to introduce the rebalance of Sage job skills. This change affects Renewal-only.

On official servers this came along with rebalances of 1st, 2nd jobs and transclass too. I am working in additional PRs for the remaining 2-2, including bard/dancer, being in separate PRs in order to keep those PRs in a reasonable size.

The implementation in this PR is based on kRO and kRO zero patch notes, iRO Wiki, rAthena and divine pride info, along with some in-game testing. I can't say everything is 100% accurate because there were discrepancies between different sources, and I could not test everything in kRO, but should be quite close.

Since clients before 2018-10 could only support up to 7 skills in Auto Spell, I added a define that allows people in Renewal to load pre-re autospell db and use the old skill list (in a mix of new/old behavior as documented in general.h) in case they can't migrate now.

I won't list all the rebalance changes in the PR description, but it may be checked in each commit text. Also, the commits are in the same order as they appear in the references below.

**Affected skills**
- SA_VOLCANO
- SA_DELUGE 
- SA_VIOLENTGALE 
- SA_FLAMELAUNCHER
- SA_FROSTWEAPON
- SA_LIGHTNINGLOADER
- SA_SEISMICWEAPON
- SA_AUTOSPELL 

References:
- [kRO 2018.10.31 patch note](https://ro.gnjoy.com/news/notice/View.asp?BBSMode=10001&seq=7033&curpage=21)
- [kRO 2018.11.28 patch note / translated](https://board.herc.ws/topic/17043-kro-patch-2018-11-28/)
- [Divine Pride](https://www.divine-pride.net/forum/index.php?/topic/3453-kro-mass-skills-balance-1st-2nd-and-transcendent-classes-skills/)

---

Since AutoSpell is tricky to test, I made some sort of "unit tests" for Auto Spell in a separate branch to try to cover as many cases. You can check the testing code in this PR in my fork: https://github.com/guilherme-gm/Hercules/pull/9/files#diff-57e0249af7272bf23858146fc95e4674d48e2ffdf89e0824333cb54e232dd56d 

It is kinda hackish, but this testing PR only extracts the functions to smaller ones with the same ones as in this main PR so I can test it using a plugin to run as many cases as possible. The same plugin can be used in this PR and it should pass for both PRE-RE and RE.

**Issues addressed:** 
Part of #2727 

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/skill.c [hunk 1]
```diff
@@ -17930,8 +17895,115 @@ static void skill_weaponrefine(struct map_session_data *sd, int idx)
...
+ * 
+ * @param list pointer to receive the list's pointer (pointer to an array)
+ * @returns list length
+ */
+static int skill_autospell_list(const struct autospell_skill **list)
+{
+	nullpo_retr(0, list);
+
+	// MUST be sorted by autospell_level (first field)
+	static const struct autospell_skill skills_list[] = {
```
Comment (guilherme-gm) [inline src/map/skill.c:17915]: I am really unsure whether this was the best approach for it. Feel free to give suggestions and discuss about it.

As I mentioned in the PR description, I did some sort of automated test to check if it still works, so if we have a better idea, we can validate it without too much effort.

Comment (MrKeiKun) [reply src/map/skill.c:17915]: de-hardcode it... move to a file within `db/` as best approach

Comment (guilherme-gm) [reply src/map/skill.c:17915]: I made a separate PR with the refactor + de-hardcode :) #3237 . Thanks!

**Comment by violent01** [comment]
There's a small problem with Hindsight. I discovered this yesterday upon implementing this draft on my fork. This is what's happening.
 Use Hindsight-> Select Heavens Drive or Earth Spike-> Recast Hindsight Now select any spells beside those two->Recast Hindsight again and choose Heavens Drive or Earth Spike again-> Hindsight will not auto-cast Heavens Drive or Earth spike during this time but instead it will auto-cast the last spell you selected instead.
Finally, you need to wait for the hindsight buff to end and reuse it for the Heavens drive or Earth spike to work again.

**Comment by guilherme-gm** [comment]
> There's a small problem with Hindsight. I discovered this yesterday upon implementing this draft on my fork. This is what's happening. Use Hindsight-> Select Heavens Drive or Earth Spike-> Recast Hindsight Now select any spells beside those two->Recast Hindsight again and choose Heavens Drive or Earth Spike again-> Hindsight will not auto-cast Heavens Drive or Earth spike during this time but instead it will auto-cast the last spell you selected instead. Finally, you need to wait for the hindsight buff to end and reuse it for the Heavens drive or Earth spike to work again.

Thank you! There was an issue created by #3237  that ended up breaking here. I made a separate PR but also included the commit here (last commit in this PR).

Thanks for testing :D

**Comment by violent01** [comment]
No worries! So far all of your rebalance changes up to MO_BLADESTOP has been nothing but fantastic. I'll keep testing every single one of your rebalance changes and I'll let you know again if I see something broken.

**Review by MishimaHaruna — APPROVED** [review]

**Pull request merged**